src/tests/testClpMap.cc

```diff
@@ -118,17 +118,51 @@ testClpMap::testSetMemLimit()
     TestMap m(2048);
     // overflow the map with entries to make sure it has lots of entries to purge below
     fillMapWithElements(m, 10);
-    const auto testEntriesBefore = m.entries();
+    auto testEntriesBefore = m.entries();
     CPPUNIT_ASSERT(testEntriesBefore > 0);
 
     m.setMemLimit(m.memoryUsed() / 2);
-    const auto entriesAfterPurge = m.entries();
+    auto entriesAfterPurge = m.entries();
     CPPUNIT_ASSERT(testEntriesBefore > entriesAfterPurge);
 
     m.setMemLimit(m.memLimit() * 2);
     // overflow the map with entries again to make sure it can grow after purging
     fillMapWithElements(m, 10);
     CPPUNIT_ASSERT(entriesAfterPurge < m.entries());
+
+    // progressively shrink to size 0, checking that the number of
+    // entries decreases each time
+    testEntriesBefore = m.entries();
+    const auto singleEntrySizeEstimate = *TestMap::MemoryCountedFor("0", 0) + 10;
+    while (m.memoryUsed() > singleEntrySizeEstimate) {
+        m.setMemLimit(m.memoryUsed() / 2);
+        entriesAfterPurge = m.entries();
+        CPPUNIT_ASSERT(testEntriesBefore > entriesAfterPurge);
+        CPPUNIT_ASSERT(entriesAfterPurge > 0);
+        testEntriesBefore = entriesAfterPurge;
+    }
+
+    // set size to 0, all entries should have been removed
+    m.setMemLimit(0);
+    CPPUNIT_ASSERT_EQUAL(size_t(0), m.entries());
+
+    // test that memory used decreases when shrinking
+    size_t curLimit = 10240;
+    size_t lastMemoryUsed;
```

Comment:
Please avoid undefined variables, especially when that does not buy you anything. To fix this, please do not just move this declaration lower, where we call m.memoryUsed() before the loop. Instead, move this variable _inside_ the loop, adjusting as needed. This change will simplify and otherwise improve this code quite a bit!